Adding documentation for an apt buildpack issue. (#2577)
18F/cg-site: _kbarticles/2024-12-09-apt-buildpack-cache-issue.md

---
layout: post
title: "Apt buildpack cache issue"
date: December 9, 2024
excerpt: Apt buildpack cache issue that can affect application deployment
---

> ## What does the buildpack cache do?

Reviewer: ```suggestion
### What does the buildpack cache do?
```

The buildpack cache (aka "app cache," or "build artifacts cache") is a per-application cache that is stored in the blob (binary large object) store and recovered when an existing application is restaged (regardless of application code or buildpack changes).

> Different buildpacks use this cache for different things. System buildpacks are cached on the cell hosting the application container and bind mounted read-only into each staging container. Buildpack URLs (to a git repo or a zip file) are downloaded every time application staging occurs.

Reviewer: ```suggestion
Different buildpacks use this cache for different things. [System buildpacks]({% link _docs/deployment/frameworks.md %}#supported-languages-and-frameworks) are cached on the VM hosting the application container and then mounted as read-only volumes into each staging container. If your application references a custom buildpack or buildpack version via a URLs (to a git repo or a zip file), those buildpacks are downloaded every time application staging occurs.
```

> ## The potential issue with the apt buildpack

Reviewer: ```suggestion
### The potential issue with the apt buildpack
```

> ### The potential issue with the apt buildpack

Reviewer: ```suggestion
### Potential cache issue with apt-buildpack 
```

> The potential issue involves the / directory disk allocation. The apt buildpack uses /tmp like other buildpacks. The / directory is allocated approximately 9% of the disk allocation. For most buildpacks, this is sufficient as all the possible packages that they could be installing is known with the only variable being the application code. The apt buildpack however, can install a vast number of packages including large packages for example the AWS CLI. When this occurs this allotted 9% of disk space is not sufficient.

Reviewer: ```suggestion
The potential issue involves the `/` directory disk allocation. The apt buildpack uses `/tmp` like other buildpacks. The `/` directory is allocated approximately 9% of the disk allocation. For most buildpacks, this is sufficient as all the possible packages that they could be installing is known with the only variable being the application code. The apt buildpack however, can install a vast number of large packages, suchas  the AWS CLI. Depending on how many packages are installed and their individual size, the allotted 9% of disk space may not be sufficient.
```

As written, I don't know if I understand. Is the apt buildpack using the `/` or `/tmp` directory? The text suggests that `/tmp` is used but then describes a problem with the `/` directory. Is it because `/tmp` is a child directory of `/`?

Author: Yes, it is because /tmp is a child directory of /.

> ## How to address the issue

Reviewer: ```suggestion
### How to address the issue
```

> The potential issue involves the `/` directory disk allocation. The apt buildpack uses `/tmp` like other buildpacks. The `/` directory is allocated approximately 9% of the disk allocation. For most buildpacks, this is sufficient as all the possible packages that they could be installing is known with the only variable being the application code. The apt buildpack however, can install a vast number of large packages, suchas  the AWS CLI. Depending on how many packages are installed and their individual size, the allotted 9% of disk space may not be sufficient.

Reviewer: ```suggestion
The potential issue with the `apt-buildpack` cache involves the `/` directory disk allocation. The `apt-buildpack` uses `/tmp` for its cache like other buildpacks. In all buildpacks, the `/` directory is allocated approximately 9% of the disk allocation, meaning that the size of the `/tmp` subdirectory cannot exceed that disk allocation. For most buildpacks, this disk allocation is sufficient as all the possible packages that they could be installing is known with the only variable being the application code. However, the `apt-buildpack` can install an unknown number of packages of variable size, such as the AWS CLI. Depending on how many packages are installed and their individual size, the allotted 9% of disk space may not be sufficient, causing application staging to fail.
```

Do we have an example of the error that customers might see?

Author: `stdout: No space left on device. /usr/bin/tar: /tmp/output-cache: Wrote only 2048 of 10240 bytes`

Reviewer: Could we add that to this documenation?

Author: There's another example on this ticket as well: https://cloud-gov-new.zendesk.com/agent/tickets/9680.

---
layout: post

> title: "Apt buildpack cache issue"

Reviewer: ```suggestion
title: "Potential buildpack cache issue with apt-buildpack"
```

date: December 9, 2024

> excerpt: Apt buildpack cache issue that can affect application deployment

Reviewer: ```suggestion
excerpt: Potential buildpack cache issue with the apt-buildpack that can cause application staging to fail
```

---

### What does the buildpack cache do?

The buildpack cache (aka "app cache," or "build artifacts cache") is a per-application cache that is stored in the blob (binary large object) store and recovered when an existing application is restaged (regardless of application code or buildpack changes).

Different buildpacks use this cache for different things. [System buildpacks]({% link _docs/deployment/frameworks.md %}#supported-languages-and-frameworks) are cached on the VM hosting the application container and then mounted as read-only volumes into each staging container. If your application references a custom buildpack or buildpack version via a URLs (to a git repo or a zip file), those buildpacks are downloaded every time application staging occurs.

### Potential cache issue with apt-buildpack 

> The potential issue with the `apt-buildpack` cache involves the `/` directory disk allocation. The `apt-buildpack` uses `/tmp` for its cache like other buildpacks. In all buildpacks, the `/` directory is allocated approximately 9% of the disk allocation, meaning that the size of the `/tmp` subdirectory cannot exceed that disk allocation. For most buildpacks, this disk allocation is sufficient as all the possible packages that they could be installing is known with the only variable being the application code. However, the `apt-buildpack` can install an unknown number of packages of variable size, such as the AWS CLI. Depending on how many packages are installed and their individual size, the allotted 9% of disk space may not be sufficient, causing application staging to fail.

Reviewer: ```suggestion
The potential issue with the `apt-buildpack` cache involves the `/` directory disk allocation. The `apt-buildpack` uses `/tmp` for its cache like other buildpacks. In all buildpacks, the `/` directory is allocated approximately 9% of the disk allocation, meaning that the size of the `/tmp` subdirectory cannot exceed that disk allocation. For most buildpacks, this disk allocation is sufficient as all the possible packages that they could be installing is known with the only variable being the application code. However, the `apt-buildpack` can install an unknown number of packages of variable size, such as the AWS CLI. Depending on how many packages are installed and their individual size, the allotted 9% of disk space may not be sufficient, causing application staging to fail with an error like:
```
